Tasks/Gradle/CodeAnalysis/SonarQube/metrics.ts

```diff
@@ -98,51 +172,84 @@ export class SonarQubeMetrics {
+                    }
+
+                });
+            });
+    }
+
+    /**
      * Repeatedly query the server to determine if the task has finished.
      * @param timeout    (optional) Time, in seconds, to wait for the analysis job to finish before returning the promise as false.
      * @param delay      (optional) Time, in seconds, to wait between polling attempts. Warning: Does not also wait for previous request to finish.
      * @returns A promise resolving true if the task finished or rejecting if there was an error or the timeout was exceeded.
      */
     private waitForTaskCompletion(timeout?:number, delay?:number):Q.Promise<boolean> {
-        tl.debug(`[SQ] Waiting for SonarQube analysis to complete `);
+        tl.debug(`[SQ] Waiting for SonarQube analysis to complete.`);
 
         // If we have previously waited, then we can immediately return true.
         if (this.analysisComplete) {
-            tl.debug(`[SQ] Analysis complete check already performed `);
+            tl.debug(`[SQ] Analysis already cached as complete.`);
             return Q.when(true);
         }
 
+        // Call isTaskComplete immediately
+        return this.isTaskComplete()
+            .then((isDone:boolean) => {
+                // If done on the first try, return fast
+                if (isDone) {
+                    return isDone;
+                }
+
+                // Otherwise, setup the delayed/repeating wait task
+                return this.setupTaskCompleteWait(timeout, delay);
+            })
+            .then((isDone:boolean) => { // This block required to avoid Promise<Promise<boolean>> return type
+                return isDone;
+            });
+    }
+
+    /**
+     * Implementation of the system to repeatedly query the server until the task has completed.
+     * Because of a detail of the implementation, the minimum time to wait is ${delay} seconds.
+     * @param timeout    (optional) Time, in seconds, to wait for the analysis job to finish before returning the promise as false.
+     * @param delay      (optional) Time, in seconds, to wait between polling attempts. Warning: Does not also wait for previous request to finish.
+     * @returns A promise resolving true if the task finished or rejecting if there was an error or the timeout was exceeded.
+     */
+    private setupTaskCompleteWait(timeout?:number, delay?:number):Q.Promise<boolean> {
         var defer = Q.defer<boolean>();
 
         // Default values
         var timeout = timeout || this.timeout;
-        var delay = delay || this.delay;
+        var delay = delay || this.delay;// Every [delay] seconds, call isTaskComplete()
 
-        // Every [delay] seconds, call isAnalysisComplete()
         var intervalTask = setInterval(() => {
             this.isTaskComplete()
                 .then((isDone:boolean) => {
                     if (isDone) {
-                        tl.debug(`[SQ] Analysis complete`);
```

Comment:
Code extracted into a method to improve readability. Cache logic that was previously here moved elsewhere to reduce redundancy and reduce the chances of not setting the cache correctly in certain circumstances.